Add EIP: Multi-block Access List Warming (#11776)
ethereum/EIPs: EIPS/eip-8289.md

---
eip: 8289
title: Multi-block Access List Warming
description: Pre-warm the per-transaction access list with items from the previous 256 block access lists
author: Toni Wahrstätter (@nerolation)

> discussions-to: https://ethereum-magicians.org/t/eip-9999-multi-block-access-list-warming/0

Reviewer: Please change the discussions-to to the correct URL (it now 404s)

status: Draft
type: Standards Track
category: Core
created: 2026-05-03

> requires: 2929, 2930, 2935, 7928

Reviewer: ```suggestion
requires: 2929, 2930, 2935, 3651, 7702, 7928
```

These two EIPs should be added as requirement (especially EIP-3651) since this alters the initial state of the access list. However I think if the initialization is worded differently (initialize access list as normal and then add the WAM items) then these do not have to be added.

This EIP defines a chain-state **warm-access multiset (WAM)** that records every item (an account address or `(address, storage_key)` pair) present in the Block Access Lists (BALs, [EIP-7928](./eip-7928.md)) of the last 256 blocks. Items in the WAM at the start of a block are treated as already-accessed for [EIP-2929](./eip-2929.md) pricing in every transaction of that block. The WAM is maintained incrementally each block: `+1` for every item in the new BAL, `-1` for every item in the BAL leaving the window. Membership tests are `O(1)`; per-block updates are `O(|BAL|)`.

## Motivation

[EIP-2929](./eip-2929.md) charges a cold-access surcharge of ~2100 gas per storage slot and ~2600 per account address the first time each transaction touches a given item, then resets the list at every new transaction. On Ethereum mainnet this surcharge accounts for a median of **17 % of every block's gas**: gas paid only to mark state as known to the EVM.

> Across blocks, the same items are touched repeatedly. Empirical measurement on 5731 mainnet blocks shows that **79 % of cold-charged items in a typical block were also touched in the previous 256 blocks** (~51 minutes). Treating those items as already-accessed at the start of the block would eliminate the surcharge they paid, saving **~14 % of total block gas at the median**.

Reviewer: ```suggestion
Across blocks, the same items are touched repeatedly. Empirical measurement on 5 731 mainnet blocks shows that **79 % of cold-charged items in a typical block were also touched in the previous 256 blocks** (~51 minutes). Treating those items as already-accessed at the start of the block would eliminate the surcharge they paid, saving **~14 % of total block gas at the median**.
```

(consistency through the doc)

### Commitment

The WAM is committed to by a binary Sparse Merkle Tree (SMT) of depth 256:

```
leaf_key(item) = SHA256(serialize(item))      # 256-bit
leaf_value     = u32 counter (0 if absent)
WAM_ROOT       = SMT root over { leaf_key → leaf_value }
```

SHA-256 is used for both leaf keys and node hashing (matching beacon-chain SSZ Merkleization; precompile `0x02`). `WAM_ROOT` is added to the block header as a new 32-byte field `wam_root`:

```python
class Header:
    ...
    wam_root: Hash32
```

The per-block transition updates the SMT incrementally: one leaf update per item in `ADD ∪ DEL`. Order does not affect the root since SMT structure depends only on leaf keys.

Inclusion and non-inclusion proofs are compressed: they carry only the ~`log2(|WAM|)` non-default siblings (empty subtrees hash to known constants), and recompute the same `WAM_ROOT`. A future ZK-friendly hash precompile (e.g., Poseidon) can replace `SHA256` without changing the structure.

### Access-list initialization in transactions

> At the start of every transaction in block `B`, the per-tx access list is initialized as today (precompiles, `tx.from`, `tx.to`, coinbase per [EIP-3651](./eip-3651.md), [EIP-2930](./eip-2930.md) access list, [EIP-7702](./eip-7702.md) authority list) **plus** every item with `WAM[item] > 0`.

Reviewer: "as today", this is not future-proof if we revisit this EIP in 3 forks.

```suggestion
At the start of every transaction in block `B`, the per-tx access list is initialized as: precompiles, `tx.from`, `tx.to`, coinbase per [EIP-3651](./eip-3651.md), [EIP-2930](./eip-2930.md) access list, [EIP-7702](./eip-7702.md) authority list **plus** every item with `WAM[item] > 0`.
```

I think this can be simplified by stating that the access list is initialized as it would be without this EIP, and every item with `WAM[item] > 0` is added as well to the list. This prevents confusion and is also future-proof in case something else in the access list changes. 

Note that currently as stated, if `tx.to` is an EIP-7702 delegated account, this is not added to the access list (while it should be)

|---:|---:|---:|---:|---:|
| 8 | 96 s | 60 % | 10.1 % | 0.6 |
| 32 | 6.4 min | 70 % | 12.0 % | 2 |
| 128 | 25.6 min | 78 % | 13.3 % | 6 |
| **256** | **51 min** | **81 %** | **13.9 %** | **10** |
| 512 | 102 min | 84 % | 14.3 % | 18 |
| 1024 | 3.4 h | 85 % | 14.5 % | 31 |
| Asymptote | n/a | 100 % | 17.5 % | n/a |

W=256 captures ~79 % of the gas asymptote at ~10 MB of WAM state and sits 4× past the finality horizon. The next doubling (W=512) triples the marginal cost per added warm conversion for only ~0.4 percentage points more, so W=256 is at the inflection of the cost-benefit curve.

Implementations that prioritise memory can pick W=8 (~0.6 MB, 60 % ops), W=16 (~1 MB, 65 %), or W=32 (~2 MB, 70 %, also one epoch). Implementations with memory to spare can pick W=512 (~18 MB, 84 %); past that the cost-benefit deteriorates sharply.

### Reuse of EIP-7928 BALs

[EIP-7928](./eip-7928.md) introduces a canonical Block Access List committed to the block header. This specification consumes that artifact directly as the source of `ADD` and `DEL` items. Without [EIP-7928](./eip-7928.md), this EIP would need an independent access-list commitment mechanism, which is out of scope.

> ### Binary SMT (not MPT) and SHA-256 (not Keccak)

Reviewer: ```suggestion
### Binary SMT (not Merkle Patricia Trie (MPT)) and SHA-256 (not Keccak)
```
